File: packages/javascript/lib/string/split.ts
export function split(str: string, ...splitter: any[]): string[];

export function split(str: string, ...splitter: Array<string>): string[];

export function split(str: string, ...splitter: Array<string> | string[]): string[] {
	if (!str) {
		return [];
	}

	const splitters = splitter.length > 0 && typeof splitter[0] === 'object' ? splitter[0] : splitter;
	const splitterLength: number = splitters.length;
	let charPattern = '';
	let strPattern = '';

	for (let i = 0; i < splitterLength; i += 1) {
		const spl = splitters[i];

		if (spl.length > 1) {
			strPattern += `${strPattern.length < 1 ? '' : '|'}${spl
				.replace(/\\/g, '\\\\')
				.replace(/\[/g, '\\[')
				.replace(/]/g, '\\]')
				.replace(/\?/g, '\\?')
				.replace(/\./g, '\\.')
				.replace(/\{/g, '\\{')
				.replace(/}/g, '\\}')
				.replace(/\+/g, '\\+')}`;
		} else if (spl === '-' || spl === '[' || spl === ']') {
			charPattern += `\\${spl}`;
		} else {
			charPattern += spl;
		}
	}

	if (charPattern.length < 1 && strPattern.length < 1) {
		return [str];
	}

	if (charPattern.length > 0) {
		charPattern = `[${charPattern}]`;
		if (strPattern.length > 0) {
			strPattern = `|${strPattern}`;
		}
	}

	return str.split(new RegExp(`${charPattern}${strPattern}+`, 'gi'));
}

File: packages/python/qsu/string/split.py
import re


def _escape_multi(spl: str) -> str:
	return (
		spl.replace('\\', '\\\\')
		.replace('[', '\\[')
		.replace(']', '\\]')
		.replace('?', '\\?')
		.replace('.', '\\.')
		.replace('{', '\\{')
		.replace('}', '\\}')
		.replace('+', '\\+')
	)


def split(str: str, *splitter) -> list:
	if not str:
		return []

	if len(splitter) > 0 and isinstance(splitter[0], (list, tuple)):
		splitters = splitter[0]
	else:
		splitters = splitter

	charPattern = ''
	strPattern = ''

	for spl in splitters:
		if len(spl) > 1:
			strPattern += ('' if len(strPattern) < 1 else '|') + _escape_multi(spl)
		elif spl in ('-', '[', ']'):
			charPattern += '\\' + spl
		else:
			charPattern += spl

	if len(charPattern) < 1 and len(strPattern) < 1:
		return [str]

	if len(charPattern) > 0:
		charPattern = '[' + charPattern + ']'
		if len(strPattern) > 0:
			strPattern = '|' + strPattern

	return re.split(charPattern + strPattern + '+', str, flags=re.IGNORECASE)
